Integrated admin dashboard

tanwarAalok opened the pull request:
## Description

Issue #692
 
Added the Admin Catalog API into the Admin Dashboard and connected it through the dashboard route. Also, cleaned the Protected Route component.

- interagrated catalog api in Admin dashboard
- added max-height in `add image` button in admin dashboard
- removed search icon from dashboard which is not visible
- integrated catalog api in `Demo.jsx` component in home page
- added toast messages in `Demo.jsx` & `Catalog.jsx` component
- cleaned `ProtectedRoute.jsx` component
- added checks before uploading/updating logo in s3 bucket
- updated failing test-cases in `demo.test.jsx`, `addlogo.js`

## What type of PR is this? (Check all applicable)

- [x] 🍕 Feature
- [x] 👨‍💻 Code Refactor
- [x] ✅ Test

## Screenshots (if applicable)

https://github.com/user-attachments/assets/b7d006b2-8903-44fe-bc5b-e797d5fbd0f6

### Test Passed

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/d78be6f2-d1ba-455f-ae72-6ac72484529a" />

### Screenshots of coverage

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/aab8ab3f-0d74-4f2e-a91d-120325027584" />

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/4b5c85c7-8a51-4158-ad27-8798ac3d39c9" />

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/cacc65e1-b7c5-4958-b36c-e222c2bd11b1" />


## Checklist
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works  
- [x] New and existing unit tests pass locally with my changes

ShlokMane committed:
Fix Add Image Button

RajeshPhayde committed:
feat: integrate GET /catalog/logos API (tests pending)

RajeshPhayde committed:
Merge remote-tracking branch 'origin/stage' into 692-integrate-admin-catalog-api

ShlokMane committed:
Merge remote-tracking branch 'origin/stage' into 692-integrate-admin-catalog-api

ShlokMane committed:
Fix Image Upload Modal

RajeshPhayde committed:
Merge remote-tracking branch 'origin/692-integrate-admin-catalog-api' into 692-integrate-admin-catalog-api

RajeshPhayde committed:
Add pagination support with skip and limit query parameters

RajeshPhayde committed:
Add companyUri input and state management to ImageUploadModal

RajeshPhayde committed:
Update CatalogItem component to use formatted dates

RajeshPhayde committed:
feat: integrate image upload functionality

RajeshPhayde committed:
refactor: enhance useApi hook to support dynamic config overrides

RajeshPhayde committed:
Merge branch '692-integrate-admin-catalog-api' of https://github.com/TeamShiksha/openlogo into 692-integrate-admin-catalog-api

tanwarAalok committed:
Merge branch 'stage' into 692-integrate-admin-catalog-api

tanwarAalok committed:
admin dashboard integrated

tanwarAalok committed:
reduce constant data

tanwarAalok committed:
sonar fixes

TanishqSingla commented on packages/app/controllers/catalog.js:
```diff
@@ -147,18 +152,10 @@ async function updateCatalogController(req, res, next) {
-        message: Messages.NAME_AND_EXT_SAME,
-      });
-    }
+    const companyName = Imagename;
```
what's the purpose of this variable?

TanishqSingla commented on packages/app/services/images.js:
```diff
@@ -24,14 +24,15 @@ class ImageServices {
+    let domainName = company.company_name;
     if (checkDb) {
-      const image = await this.imageRepository.fetchImage(company);
+      const image = await this.imageRepository.fetchImage(company.company_name);
```
```suggestion
      const image = await this.imageRepository.fetchImage(domainName);
```

TanishqSingla commented on packages/app/services/images.js:
```diff
@@ -57,7 +58,7 @@ class ImageServices {
     const dataList = [];
     for (const company of companyList) {
```
Can we use `Promise.all` or `Promise.allSettled` here?

TanishqSingla commented on packages/ui/__tests__/components/Catalog.test.jsx:
```diff
@@ -16,13 +83,15 @@ describe("Catalog Component", () => {
 
   it("Add image button should be visible", () => {
     render(<Catalog />);
+    vi.runAllTimers();
```
`runAllTimers` is mentioned in `beforeEach`, do we need this after the render?

Allan2000-Git replied:
Yes, not required after render. Have removed all the instances

TanishqSingla commented on packages/app/services/images.js:
```diff
@@ -24,14 +24,15 @@ class ImageServices {
 
-    const imageUrl = `${default_extension}/${domainName}`;
+    const version = new Date(company.updated_at).getTime();
+    const imageUrl = `${default_extension}/${domainName}?v=${version}`;
```
nit: This should be done in a static function containing tests.

Allan2000-Git replied:
There are no tests containing this line of code

TanishqSingla commented on packages/app/services/images.js:
```diff
@@ -102,15 +103,21 @@ class ImageServices {
   }
 
   async updateImageById(id, updateObj) {
```
please add JSDoc for this

Allan2000-Git replied:
have added the jsdoc

TanishqSingla commented on packages/app/services/images.js:
```diff
@@ -102,15 +103,21 @@ class ImageServices {
 
-  async getImagesByUserId(userId) {
-    return await this.imageRepository.getAllImageByUserId(userId);
+  async getImagesByUserId(userId, skip, limit) {
```
Same as above

Allan2000-Git replied:
have added the jsdoc

TanishqSingla commented on packages/ui/__tests__/components/Catalog.test.jsx:
```diff
@@ -1,11 +1,78 @@
+      return { data, loading, errorMsg, makeRequest };
+    },
+  };
+});
```
We should not write logic in tests, this is way too complicated.

tanwarAalok replied:
I am not sure about this too, I tried many things before I could make the tests run successfully.

Allan2000-Git replied:
I have simplified it using the mock data

TanishqSingla commented on packages/ui/__tests__/components/Catalog.test.jsx:
```diff
@@ -35,160 +104,187 @@ describe("Catalog Component", () => {
-      const companyImage = screen.getByText(company.companyImage);
-      expect(companyImage).toBeInTheDocument();
+      expect(
+        screen.getByText(
```
`getByText` is itself an assertion, so `expect` here is redundant.

TanishqSingla commented on packages/ui/__tests__/components/Catalog.test.jsx:
```diff
@@ -35,160 +104,187 @@ describe("Catalog Component", () => {
+        screen.queryByText(
+          `${COMPANIES[10].company_name.toLowerCase()}.${COMPANIES[10].extension}`
+        )
+      ).not.toBeInTheDocument();
```
`queryByText` returns `null` if the element is not present, so `.not.null()` would be much more accurate

Allan2000-Git replied:
I have used .not.null() and tested the same

TanishqSingla commented on packages/ui/__tests__/components/Catalog.test.jsx:
```diff
@@ -35,160 +104,187 @@ describe("Catalog Component", () => {
+    vi.runAllTimers();
+
+    expect(screen.getByText("amazon.png")).toBeInTheDocument();
+    expect(screen.queryByText("apple.png")).not.toBeInTheDocument();
```
Same as above `getBy*` are assertions, so `expect - .toBeInDocument` are redundant

TanishqSingla commented on packages/app/repositories/images.js:
```diff
@@ -51,8 +51,19 @@ class ImagesRepository extends BaseRepository {
+    const images = await this.model
+      .find({ user_id: userId })
+      .skip(skip)
+      .limit(limit);
```
Count query can sometimes be slow, both these queries can be ran in paralllel using `Promise.all`

TanishqSingla commented on packages/ui/src/utils/ProtectedRoute.jsx:
```diff
@@ -1,54 +1,21 @@
     return <Navigate to="/" replace state={{ from: location.pathname }} />;
   }
 
-  if (!adminOnly) {
```
is removing `adminOnly` part of the issue itself?

tanwarAalok replied:
We don't need admin logic now in protected route as we are providing dropdown in dashboard itself. There is no additional /admin route. So this part removed.

TanishqSingla commented on packages/ui/src/components/common/modal/Modal.jsx:
```diff
@@ -35,8 +35,13 @@ const Modal = ({
+      className={styles["modal-overlay"]}
+      onClick={handleOverlayClick}
+    >
+      <dialog
```
html dialog itself renders on a top layer, this will likely render on top of overlay, can you confirm the functionality before is same as after the changes?

TanishqSingla commented on packages/ui/src/components/catalog/Catalog.jsx:
```diff
@@ -1,53 +1,114 @@
+    setUpdateImageId(id);
+  };
+
+  const filteredCompanies =
```
better to have this function wrapped inside `useMemo` so that we don't recompute it every time

TanishqSingla commented on packages/ui/src/hooks/useApi.js:
```diff
@@ -15,13 +15,15 @@ export const useApi = (config) => {
   const [loading, setLoading] = useState(false);
 
-  const makeRequest = async () => {
+  const makeRequest = async (dynamicConfig = {}) => {
```
Have you evaluated your scenario and compared other components that justifies the need for this?

tanwarAalok replied:
![image](https://github.com/user-attachments/assets/23835649-a21a-4779-8286-a6a0a734d62b)

It would help to pass body data in POST request or update request.

tanwarAalok committed:
code-review-changes

tanwarAalok committed:
Merge branch 'stage' into 692-integrate-admin-catalog-api

aps08 commented:
Aalok is not contributing, so @TanishqSingla if you are free then you can get started on this.

Allan2000-Git commented:
@TanishqSingla It would be helpful to add more detailed comments, as the current ones are not very descriptive.

aps08 commented:
@Allan2000-Git @TanishqSingla what is the update on this ?

Allan2000-Git committed:
Merge branch 'stage' into 692-integrate-admin-catalog-api

Allan2000-Git committed:
Merge branch 'stage' into 692-integrate-admin-catalog-api

Allan2000-Git committed:
refactor: catalog tests and mock api response + reset company id reupload modal close

Allan2000-Git committed:
fix: added jsdoc image services & used .not.null for queryByText

aps08 commented on packages/app/services/images.js:
```diff
@@ -55,18 +56,27 @@ class ImageServices {
+      companyList.map(async (company) => {
+        const signedUrl = await this.fetchImageByCompanyFree(
+          company,
+          undefined,
```
why passing undefined ?

theboyofdream committed:
Merge branch 'stage' of https://github.com/TeamShiksha/openlogo into 692-integrate-admin-catalog-api

theboyofdream committed:
Merge branch 'stage' of https://github.com/TeamShiksha/openlogo into 692-integrate-admin-catalog-api

theboyofdream committed:
demo component & catalog component
- added folder-name `imageUrl` const in `fetchImageByCompanyFree` function in images.js
- added checks before uploading/updating logo in s3 bucket
- updated failing test-cases in `demo.test.jsx`, `addlogo.js`
- removed search icon from dashboard which is not visible
- added max-height in `add image` button
- added toast messages for image upload/reupload errors in dashboard admin

theboyofdream committed:
sonar cloud fix

TanishqSingla commented on packages/ui/__tests__/components/Demo.test.jsx:
```diff
@@ -37,68 +27,63 @@ describe("Demo Component", () => {
-        { companyName: "Airbnb", image: "airbnb-logo.svg" },
-      ],
+      data: {
+        data: [
+          { companyName: "Google", image: "google-logo.svg" },
+          { companyName: "Facebook", image: "facebook-logo.svg" },
+          { companyName: "Zoom", image: "zoom-logo.svg" },
+          { companyName: "DoorDash", image: "doordash-logo.svg" },
```
I get the `data` change, but why are the values also changed here, this will just introduce the extra work of changing values in each test

TanishqSingla commented on packages/ui/src/utils/Constants.js:
```diff
@@ -780,3 +751,10 @@ export const CHANGE_PASSWORD_FIELDS = [
 };
+
+export const MOCK_DEMO_COMPANIES = [
+  { companyName: "Google", image: "google-logo.svg" },
+  { companyName: "Facebook", image: "facebook-logo.svg" },
+  { companyName: "Zoom", image: "zoom-logo.svg" },
```
Why do we have mock values in utils? Is this used in tests?

TanishqSingla commented on packages/app/services/images.js:
```diff
@@ -24,14 +24,15 @@ class ImageServices {
 
-    const imageUrl = `${default_extension}/${domainName}`;
+    const version = new Date(company.updated_at).getTime();
+    const imageUrl = `${process.env.KEY}/${default_extension}/${domainName}?v=${version}`;
```
This is a method function and you're reading a global variable `process.env.KEY`. Not a good pattern to have predictable code.
Either move this to a function or initialize `KEY` in class in constructor. That way initialized class won't produce different output.

Also can we have a better name than `KEY`, not descriptive enough of where this key is from and where it belongs.

TanishqSingla commented on packages/ui/src/components/catalog/Catalog.jsx:
```diff
@@ -1,107 +1,205 @@
+    headers: { "Content-Type": "multipart/form-data" },
+  });
+
+  useEffect(() => {
+    makeRequest();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [pageNum]);
```
is the interface going to be an infinite scroll pagination? If not then why are we not having support for url search params which will allow the user to have browser navigation controls
